src/mdns/mdns_avahi.cpp

```diff
@@ -533,34 +629,35 @@ otbrError PublisherAvahi::PublishService(const char *   aHostName,
         used = static_cast<size_t>(reinterpret_cast<uint8_t *>(curr) - reinterpret_cast<uint8_t *>(buffer));
     }
 
-    for (Services::iterator it = mServices.begin(); it != mServices.end(); ++it)
+    serviceIt = FindService(aName, aType);
+
+    if (serviceIt == mServices.end())
     {
-        if (!strncmp(it->mName, aName, sizeof(it->mName)) && !strncmp(it->mType, aType, sizeof(it->mType)) &&
-            it->mPort == aPort)
-        {
-            otbrLog(OTBR_LOG_INFO, "MDNS update service %s", aName);
-            error = avahi_entry_group_update_service_txt_strlst(
-                mGroup, AVAHI_IF_UNSPEC, mProtocol, static_cast<AvahiPublishFlags>(0), aName, aType, mDomain, last);
-            SuccessOrExit(error);
-            ret = OTBR_ERROR_NONE;
-            ExitNow();
-        }
+        SuccessOrExit(ret = CreateService(*mClient, aName, aType, serviceIt));
+    }
+    else if (serviceIt->mHostName != safeHostName || serviceIt->mPort != aPort)
+    {
+        SuccessOrExit(ret = ResetGroup(serviceIt->mGroup));
+    }
+    else
+    {
+        otbrLog(OTBR_LOG_INFO, "MDNS update service %s for host '%s'", aName, safeHostName);
+        error = avahi_entry_group_update_service_txt_strlst(serviceIt->mGroup, AVAHI_IF_UNSPEC, mProtocol,
+                                                            AvahiPublishFlags{}, aName, aType, mDomain, last);
+        ExitNow();
     }
 
-    otbrLog(OTBR_LOG_INFO, "MDNS create service %s", aName);
-    error = avahi_entry_group_add_service_strlst(mGroup, AVAHI_IF_UNSPEC, mProtocol, static_cast<AvahiPublishFlags>(0),
+    otbrLog(OTBR_LOG_INFO, "MDNS create service %s for host '%s'", aName, safeHostName);
+    error = avahi_entry_group_add_service_strlst(serviceIt->mGroup, AVAHI_IF_UNSPEC, mProtocol, AvahiPublishFlags{},
```

Comment:
Do we want to remove the service from the service list if failed to publish the service?

Reply:
I think a better practice is that we add the service to the list only after successfully publishing it. Thoughts?

Reply:
Ok, the advertising proxy seems to call the `Publish*` methods periodically, so I thought it would be better to allocate the structures once and just repeat the Avahi calls, but I can change it.